### PR TITLE
Remove FrameData and update OrderManager

### PR DESCRIPTION
**Pull request by teinarss**

### CODE REVIEW

File: OpenRA.Game/Network/OrderManager.cs
Comment (teinarss) [inline OpenRA.Game/Network/OrderManager.cs]: Maybe rename to UpdateOrdersForFrame?

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: Maybe `AddPendingPacket`?

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: > --> check if correct before we merge this

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: `GameStarted && !ClientsNotReadyForFrame().Any();` and then remove `IsReadyForFrame` below?

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [reply OpenRA.Game/Network/OrderManager.cs]: Do the next steps require this to be its own method? IMO it might be cleaner to have this loop directly inline in `OrderManager.Tick`. It looks like it could be combined with the `ProcessOrder` loop, which would remove the need for the `clientOrders` field.

**Review by pchote — COMMENTED** [review]
A few more first-look thoughts. I still need to test and try to break this.

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [reply OpenRA.Game/Network/OrderManager.cs]: Or probably better to keep this inline in `TickImmediate`, like the `immediatePackets` update.

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: If this stays as the only use of `ClientsNotReadyForFrame` then IMO it would be clearer to get rid of that and use `return activeClients.All(client => pendingPackets[client].Count > 0);` here instead.

File: OpenRA.Game/Network/OrderManager.cs
Comment (pchote) [inline OpenRA.Game/Network/OrderManager.cs]: After more digging I was finally able to resolve one of my main concerns about these changes: *we appear to be completely ignoring the frame defined in the packet header, so how on earth are these orders still being processed on the right tick?*

The answer is that
https://github.com/OpenRA/OpenRA/blob/406e15c2413064f08bafcdb94d81865191f2d46b/OpenRA.Game/Network/OrderManager.cs#L93-L95
inserts an initial set of empty packets into the queue, and we then add and process one packet per frame, meaning that our processing stays synchronised with the order latency and everything (hopefully) works as it should.

I think its a bad idea to leave this implicit: we should explain and validate this assumption, throwing an exception to guard against future PRs potentially regressing this. Combining with my suggestion above to inline the processing logic:

```suggestion
			var clientOrders = new List<ClientOrder>();
			foreach (var clientId in activeClients)
			{
				// The IsReadyForNextFrame check above guarantees that all clients have sent a packet
				var frameData = pendingPackets[clientId].Dequeue();

				// Orders are synchronised by sending an initial FramesAhead set of empty packets
				// and then making sure that we enqueue and process exactly one packet for each player each tick.
				// This may change in the future, so sanity check that the orders are for the frame we expect
				// and crash early instead of risking desyncs.
				var frameNumber = BitConverter.ToInt32(frameData, 0);
				if (frameNumber != NetFrameNumber)
					throw new InvalidDataException($"Attempted to process orders from client {clientId} for frame {frameNumber} on frame {NetFrameNumber}");

				foreach (var order in frameData.ToOrderList(World))
				{
					UnitOrders.ProcessOrder(this, World, clientId, order);
					clientOrders.Add(new ClientOrder { Client = clientId, Order = order });
				}
			}
```


We can then remove the `clientOrders` field on the class and remove the `.ToList()` when giving these to `UpdateSyncReport`.

**Comment by pchote** [comment]
Replays (local + server) and game saves seem to be working as expected, so I think this should be good to go once we address my comments above.

**Comment by teinarss** [comment]
updated and rebased.

**Review by pchote — APPROVED** [review]

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/2e106fef0877e35c9b2a31aca6bd05757eb05420...9087fcda6d587b9c9a14fdb0b192cd4649f2e025

And sorry, I forgot I noticed a silent rebase error, will file a fix myself.

**Pull request merged**
